### PR TITLE
installer: add interactive configuration process

### PR DESCRIPTION
**Pull request by mufeng**
Currently user must define a configuration file to lead the
installation process. In order to reduce the learning complexity of the
configuration items, we introduce this interactive process to guide the
user to genarate a configuration file and use that during the following
installation process.

This interactive process is triggered if no configuration file is
provided through the commandline or if interactive parameter
--interactive or -I is specifically provided through commandline.

The genarated configuration file can be reused later by passing through
the --config parameter.

Signed-off-by: Feng Mu Feng.Mu@windriver.com

### CODE REVIEW

**Comment by zeddii** [comment]
Just commenting to let you know that I see this, but have intermittent / slow internet access right now. I'll have a look at this ASAP.

File: installers/cubeit-installer [hunk 1]
```diff
@@ -72,6 +73,7 @@ cat << EOF
   cubeit-installer <base image> <device>
 
     -b: use btrfs
+    -I(--interactive): use the interactive configuration interface
```
Comment (zeddii) [inline installers/cubeit-installer:76]: -i would be better for this. We currently only use lower case, or long form parameters. -i is consistent with that.

File: installers/cubeit-installer [hunk 2]
```diff
@@ -168,8 +177,43 @@ if [ -e "${LOCAL_POST_FUNCTION_DEFS}" ] ; then
...
 
+# Check if interactive mode will be used
+INSTALL_TYPE="full"
+if [ -n "$INTERACTIVE_MODE" ] && [ "$INTERACTIVE_MODE" -eq 1 ]; then
+	debugmsg ${DEBUG_INFO} "Entering interactive mode..."
+	SAVE_CONFIG_FOLDER="saved_config"
+	recursive_mkdir ${SAVE_CONFIG_FOLDER}
+	interactive_config ${raw_dev} ${SAVE_CONFIG_FOLDER}
+	if [ $? -ne 0 ]; then
+		debugmsg ${DEBUG_CRIT} -e "\n\n\nFailed to generate one using interactive mode, run again or specify a config via --config option."
```
Comment (zeddii) [inline installers/cubeit-installer:207]: Failed to generate config using interactive mode ..

File: installers/cubeit-installer [hunk 3]
```diff
@@ -878,6 +905,9 @@ smart update"
 
 fi
 
+if [ -d "$CURRENTDIR/$PUPPETDIR" ]; then
```
Comment (zeddii) [inline installers/cubeit-installer:908]: Why is this needed ? When it wasn't needed before ?

File: sbin/cubeit [hunk 4]
```diff
@@ -33,6 +41,7 @@ cat << EOF
   options:
 
      --config <script>: use the specified configuration script found in ~/.overc/
+     -I(--interactive): use the interactive configuration interface
```
Comment (zeddii) [inline sbin/cubeit:44]: same comment, this should be "-i" or "--interactive"

File: sbin/cubeit [hunk 5]
```diff
@@ -235,6 +238,24 @@ if [ "${#ROOTFS_LABEL}" -gt 16 ]; then
 	fi
 fi
 
+# Check if interactive mode will be used
+if [ -z ${CONFIG_FILES} ] || ([ -n "$INTERACTIVE_MODE" ] && [ "$INTERACTIVE_MODE" -eq 1 ]); then
```
Comment (zeddii) [inline sbin/cubeit:242]: Why is this same check/trigger for interactive mode being run from both cubeit and cubeit-installer ?

File: sbin/cubeit [hunk 6]
```diff
@@ -500,6 +521,7 @@ IFS=$OLDIFS
 	    # create a configuration that can be read by the cubeit-installer
 	    if [ -v TARGET_CONFIG_FILE -a -n "$TARGET_CONFIG_FILE" ]; then
 		cat $TARGET_CONFIG_FILE |
+			sed "s|ARTIFACTS_DIR=.*|ARTIFACTS_DIR=${INSTALLER_TARGET_IMAGES_DIR}|" | \
```
Comment (zeddii) [inline sbin/cubeit:524]: it isn't clear why this is being changed. We are already modifying ARTIFACTS_DIR, and now we are modfying it for ARTIFACTS_DIR=.

File: sbin/cubeit [hunk 7]
```diff
@@ -527,9 +549,12 @@ IFS=$OLDIFS
 	# puppet modules etc.
 	if [ -v INSTALL_PUPPET_DIR ]; then
 	    if [ -d ${INSTALLER_FILES_DIR}/${INSTALL_PUPPET_DIR} ]; then
-		debugmsg ${DEBUG_INFO} "Copying Puppet files to install media"
-		recursive_mkdir ${mnt_rootfs}${INSTALLER_TARGET_FILES_DIR}/puppet
-		cp -r ${INSTALLER_FILES_DIR}/${INSTALL_PUPPET_DIR}/* ${mnt_rootfs}${INSTALLER_TARGET_FILES_DIR}/puppet/
+			INSTALL_PUPPET_DIR=${INSTALLER_FILES_DIR}/${INSTALL_PUPPET_DIR}
+		fi
+		if [ -d ${INSTALL_PUPPET_DIR} ]; then
```
Comment (zeddii) [inline sbin/cubeit:554]: This change doesn't make sense. The end result is the same.

File: sbin/functions.sh [hunk 8]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+
+	# Container selection
+	local filecount=0
+	local filelist="" menuitems=""
+	if [ -z "$ARTIFACTS_DIR" ]; then
+		if [ -e "tmp/deploy/images" ]; then
+			filecount=`ls tmp/deploy/images/ -1 | wc -l`
+			case "$filecount" in
+				0)
+					debugmsg ${DEBUG_CRIT}  "Project build incomplete. Please try finish the build first."
```
Comment (zeddii) [inline sbin/functions.sh:1133]: We should never reference a build in the installer. Simply indicate that images were not found.

File: sbin/functions.sh [hunk 9]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+	else
+		echo "DISTRIBUTION=\"OverC\"" >> ${localconf}
+	fi
+
+	# Container selection
+	local filecount=0
+	local filelist="" menuitems=""
+	if [ -z "$ARTIFACTS_DIR" ]; then
+		if [ -e "tmp/deploy/images" ]; then
+			filecount=`ls tmp/deploy/images/ -1 | wc -l`
```
Comment (zeddii) [inline sbin/functions.sh:1130]: We should never look in tmp/deploy/*

The installer knows nothing about bitbake, or the structure of a build. It simply takes a directory of artifacts.

File: sbin/functions.sh [hunk 10]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+			case "$filecount" in
+				0)
+					debugmsg ${DEBUG_CRIT}  "Project build incomplete. Please try finish the build first."
+					return 1
+					;;
+				1)
+					ARTIFACTS_DIR="`pwd`/tmp/deploy/images/`ls tmp/deploy/images/`"
+					;;
+				*)
+					filelist=`ls tmp/deploy/images` menuitems=""
```
Comment (zeddii) [inline sbin/functions.sh:1140]: Again. Never reference tmp/deploy/*

File: sbin/functions.sh [hunk 11]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+					for i in $filelist; do
+						menuitems="$menuitems $i $i "
+					done
+					dialog --no-cancel --menu "Found multiple machine target in your project, Please choose one:" \
```
Comment (zeddii) [inline sbin/functions.sh:1144]: Or this. The installer knows nothing about machines, or what bitbake leaves as build outputs. Take a path to artefacts and use them.

File: sbin/functions.sh [hunk 12]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+			esac
+		elif [ -e "images" ]; then
+			ARTIFACTS_DIR="`pwd`/images"
+		else
+			# default to current dir
+			ARTIFACTS_DIR="`pwd`"
+		fi
+	fi
+	echo "ARTIFACTS_DIR=\"${ARTIFACTS_DIR}\"" >> ${localconf}
+	filecount=`ls $ARTIFACTS_DIR -1 | grep -v menifest | wc -l`
```
Comment (zeddii) [inline sbin/functions.sh:1157]: "menifest" ? That should be "manifest"

File: sbin/functions.sh [hunk 13]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+	# exclude manifest files
+	filelist=`ls $ARTIFACTS_DIR | grep -v manifest` menuitems=""
+	local checklistmenu=""
+	for i in $filelist; do
+		menuitems="$menuitems $i $i "
+		checklistmenu="$checklistmenu $i $i off "
+	done
+
+	# Kernel selection
+	dialog --no-tags --menu "Please choose kernel:" \
```
Comment (zeddii) [inline sbin/functions.sh:1172]: We should have some sort of kernel regex, instead of checking for a manifest. That is a build output that we can't use.

File: sbin/functions.sh [hunk 14]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+	dialog_res=$(cat $tmpfile)
+	if [ -z "${dialog_res}" ]; then
+		debugmsg ${DEBUG_CRIT} "Kernel not specified. Please try again."
+		return 1
+	fi
+	echo "INSTALL_KERNEL=\"\${ARTIFACTS_DIR}/${dialog_res}\"" >> ${localconf}
+
+	# Rootfs selection
+	dialog --no-tags --menu "Please choose rootfs:" \
+		$(expr $filecount + 7) 100 $filecount $menuitems 2>$tmpfile
```
Comment (zeddii) [inline sbin/functions.sh:1183]: Again. It is unclear how images, kernels and other binary blobs are being categorized. Providing a list is nice, but if we are listing invalid options to the user, it isn't worth doing.

File: sbin/functions.sh [hunk 15]
```diff
@@ -964,6 +964,469 @@ clean_up()
...
+		debugmsg ${DEBUG_CRIT} "Containers not specified. Please try again."
+		return 1
+	fi
+
+	local hdcontainers=${dialog_res} fmtcontainers=""
+
+	# VT and network prime selection
+	menuitems=""
+	filecount=0
+	for i in ${hdcontainers}; do
```
Comment (zeddii) [inline sbin/functions.sh:1223]: The goal of the interactive installer's questions was that they should be scripts that are dectected, sourced and then run. 

To change these, you must change the installer, or it's library routines.

i.e. there should be something like 'prompts/foo.sh", that will be detected, sourced and then executed. These scripts can assume any of the installer variables, and also be passed information in a standard routine like "foo_installer $artefacts ..."

File: installers/cubeit-installer (hunk 3 above)
Comment (mufeng) [reply installers/cubeit-installer:908]: I want to support relative path instead of hardcoded path, so here make a test and see if it is a relative one.

File: sbin/cubeit (hunk 5 above)
Comment (mufeng) [reply sbin/cubeit:242]: We want to support interactive installer for both host cross install and on-target installer image, so I think we should support this interactive mode in both.

File: sbin/cubeit (hunk 6 above)
Comment (mufeng) [reply sbin/cubeit:524]: The previous modification was just for HDINSTALL_CONTAINERS variable because they are places under the new "$ARTIFACTS_DIR/containers" subfolder. Somehow the kernel/essential/ramfs rootfs are still placed under the $ARTIFACTS_DIR, so we need to make sure that this is correctly set too.

File: sbin/cubeit (hunk 7 above)
Comment (mufeng) [reply sbin/cubeit:554]: If $INSTALL_PUPPET_DIR is configured by interactive configure process, it will not be found in the ${INSTALLER_FILES_DIR}/${INSTALL_PUPPET_DIR}. So here we make sure that we don't depend on this INSTALLER_FILES_DIR variable.

File: sbin/functions.sh (hunk 13 above)
Comment (mufeng) [reply sbin/functions.sh:1172]: What about this regex: (._[zu]Image._)|(._vmlinu[xz]._)

File: sbin/functions.sh (hunk 14 above)
Comment (mufeng) [reply sbin/functions.sh:1183]: If we restrict the name/rootfs file type, user might be confused if they made their own image format other than our restricted one. If you think we should do this restriction, please just let me know how  we should define the rootfs type.

**Comment by zeddii** [comment]
On Sep 15, 2016 10:18 AM, "Feng Mu" notifications@github.com wrote:

> @mufeng commented on this pull request.
> 
> ---
> 
> In installers/cubeit-installer:
> 
> > @@ -878,6 +905,9 @@ smart update" fi +if [ -d "$CURRENTDIR/$PUPPETDIR"
> > ]; then
> 
> I want to support relative path instead of hardcoded path, so here make a
> test and see if it is a relative one.

OK.  This should be fine

Bruce

**Comment by zeddii** [comment]
On Sep 15, 2016 10:20 AM, "Feng Mu" notifications@github.com wrote:

> @mufeng commented on this pull request.
> 
> ---
> 
> In sbin/cubeit:
> 
> > @@ -235,6 +238,24 @@ if [ "${#ROOTFS_LABEL}" -gt 16 ]; then fi fi +#
> > Check if interactive mode will be used +if [ -z ${CONFIG_FILES} ] || ([ -n
> > "$INTERACTIVE_MODE" ] && [ "$INTERACTIVE_MODE" -eq 1 ]); then
> 
> We want to support interactive installer for both host cross install and
> on-target installer image, so I think we should support this interactive
> mode in both.

The interactive questions/answers are only in one place. The cubeit phase
should be identical.

Bruce

**Comment by zeddii** [comment]
On Sep 15, 2016 10:24 AM, "Feng Mu" notifications@github.com wrote:

> @mufeng commented on this pull request.
> 
> ---
> 
> In sbin/cubeit:
> 
> > @@ -500,6 +521,7 @@ IFS=$OLDIFS # create a configuration that can be
> > read by the cubeit-installer if [ -v TARGET_CONFIG_FILE -a -n
> > "$TARGET_CONFIG_FILE" ]; then cat $TARGET_CONFIG_FILE | + sed
> > "s|ARTIFACTS_DIR=.*|ARTIFACTS_DIR=${INSTALLER_TARGET_IMAGES_DIR}|" | \
> 
> The previous modification was just for HDINSTALL_CONTAINERS variable
> because they are places under the new "$ARTIFACTS_DIR/containers"
> subfolder. Somehow the kernel/essential/ramfs rootfs are still placed under
> the $ARTIFACTS_DIR, so we need to make sure that this is correctly set too

The tested directory is still the same in this. There's no issue using the
variable as-is. The point is that this is unnecessary churn, mixed in with
real changes

Bruce

.

**Comment by zeddii** [comment]
On Sep 15, 2016 10:38 AM, "Feng Mu" notifications@github.com wrote:

> @mufeng commented on this pull request.
> 
> ---
> 
> In sbin/functions.sh:
> 
> > - filecount=`ls $ARTIFACTS_DIR -1 | grep -v menifest | wc -l` + if [
> >   $filecount -lt 3 ]; then + debugmsg ${DEBUG_CRIT} "Artifacts are not
> >   sufficient. Please verify." + return 1 + fi + + # exclude manifest files +
> >   filelist=`ls $ARTIFACTS_DIR | grep -v manifest` menuitems="" + local
> >   checklistmenu="" + for i in $filelist; do + menuitems="$menuitems $i $i " +
> >   checklistmenu="$checklistmenu $i $i off " + done + + # Kernel selection +
> >   dialog --no-tags --menu "Please choose kernel:" \
> 
> What about this regex: (.[zu]Image.)|(.vmlinu[xz].)

That looks fine

Bruce

**Comment by zeddii** [comment]
On Sep 15, 2016 10:45 AM, "Feng Mu" notifications@github.com wrote:

> @mufeng commented on this pull request.
> 
> ---
> 
> In sbin/functions.sh:
> 
> > - checklistmenu="$checklistmenu $i $i off " + done + + # Kernel
> >   selection + dialog --no-tags --menu "Please choose kernel:" \ + $(expr
> >   $filecount + 7) 100 $filecount $menuitems 2>$tmpfile + dialog_res=$(cat
> >   $tmpfile) + if [ -z "${dialog_res}" ]; then + debugmsg ${DEBUG_CRIT}
> >   "Kernel not specified. Please try again." + return 1 + fi + echo
> >   "INSTALL_KERNEL=\"\${ARTIFACTS_DIR}/${dialog_res}\"" >> ${localconf} + + #
> >   Rootfs selection + dialog --no-tags --menu "Please choose rootfs:" \ +
> >   $(expr $filecount + 7) 100 $filecount $menuitems 2>$tmpfile
> 
> If we restrict the name/rootfs file type, user might be confused if they
> made their own image format other than our restricted one. If you think we
> should do this restriction, please just let me know how we should define
> the rootfs type.

Keep it simple for now. The kernel regex testing for bzImage/zImage/uImage
and tgz (etc) for roots options is good enough. From what I see we are just
showing all files found in the listing.. if we are listing already sorted
directories then the complete listing would be OK

Bruce
